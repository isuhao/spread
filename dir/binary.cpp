#include "binary.hpp"
#include <mangle/stream/servers/file_stream.hpp>
#include <mangle/stream/servers/outfile_stream.hpp>
#include <mangle/stream/servers/null_stream.hpp>
#include "hash/hash_stream.hpp"
#include <vector>

using namespace Spread;
using namespace std;
using namespace Mangle::Stream;

typedef Hash::DirMap DirMap;

const int MAGIC_NUMBER = 0x4d3bface;

static void fail(const std::string &msg) { throw std::runtime_error(msg); }
static void failDir(const std::string &msg)
{ fail("Error parsing directory file: " + msg); }

Hash Spread::Dir::read(Hash::DirMap &dir, StreamPtr strm)
{
  HashStream inf(strm);

  int tmp;

  // Read magic number
  inf.read(&tmp, 4);
  if(tmp != MAGIC_NUMBER)
    failDir("invalid magic number");

  // Number of hashes
  inf.read(&tmp, 4);

  // If someone throws us a malicious object, better to throw than
  // to crash.
  if(tmp > 1024*1024)
    failDir("too many directory hashes");

  Hash h;
  uint8_t hbuf[40];
  vector<char> buf;
  for(int i=0; i<tmp; i++)
    {
      // Read and parse hash
      inf.read(hbuf, 40);
      h.copy(hbuf);

      // String length + data
      int len = 0;
      inf.read(&len, 2);
      if(buf.size() < len)
        buf.resize(len);
      inf.read(&buf[0], len);

      // Store
      dir[string(&buf[0], len)] = h;
    }

  return inf.finish();
}

Hash Spread::Dir::write(const Hash::DirMap &dir, StreamPtr strm)
{
  HashStream ouf(strm);

  int tmp;

  tmp = MAGIC_NUMBER;
  ouf.write(&tmp, 4);

  tmp = dir.size();
  ouf.write(&tmp, 4);

  for(DirMap::const_iterator it = dir.begin();
      it != dir.end(); it++)
    {
      ouf.write(it->second.getData(), 40);

      const void *p = it->first.c_str();
      tmp = it->first.size();

      // Write size (2 bytes) first, then the string it self
      ouf.write(&tmp, 2);
      ouf.write(p, tmp);
    }

  return ouf.finish();
}

Hash Spread::Dir::read(Hash::DirMap &dir, const std::string &file)
{ return read(dir, FileStream::Open(file)); }

Hash Spread::Dir::write(const Hash::DirMap &dir, const std::string &file)
{ return write(dir, OutFileStream::Open(file)); }

Hash Spread::Dir::hash(const Hash::DirMap &dir)
{
  return write(dir, StreamPtr(new NullStream));
}
