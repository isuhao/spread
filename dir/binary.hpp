#ifndef __SPREAD_BINARY_DIR_HPP_
#define __SPREAD_BINARY_DIR_HPP_

#include <hash/hash.hpp>
#include <mangle/stream/stream.hpp>

namespace Spread
{
  namespace Dir
  {
    /* Read and write binary dir files. Returns the hash of the data
       read or written.
     */
    extern Hash read(Hash::DirMap &dir, const std::string &file);
    extern Hash write(const Hash::DirMap &dir, const std::string &file);

    // Stream versions
    extern Hash read(Hash::DirMap &dir, Mangle::Stream::StreamPtr strm);
    extern Hash write(const Hash::DirMap &dir, Mangle::Stream::StreamPtr strm);

    /* Return the hash of the dir, without actually writing the data
       anywhere.
     */
    extern Hash hash(const Hash::DirMap &dir);
  };
}
#endif
