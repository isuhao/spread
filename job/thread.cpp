#include "thread.hpp"

#include <boost/thread.hpp>
#include <assert.h>

using namespace Jobify;

struct ThreadObj
{
  Job *j;
  void operator()()
  {
    j->run();
    assert(j->getInfo()->isFinished());
    delete j;
  }
};

void Thread::sleep(double seconds)
{
  int msecs = (int)(seconds*1000000);
  boost::this_thread::sleep(boost::posix_time::microseconds(msecs));
}

void Thread::run(Job *j, bool async)
{
  ThreadObj to;
  to.j = j;
  if(async) boost::thread trd(to);
  else to();
}
