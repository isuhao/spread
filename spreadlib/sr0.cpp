#include "sr0.hpp"
#include "tasks/install.hpp"
#include "tasks/fileop.hpp"
#include <boost/filesystem.hpp>
#include <mangle/stream/servers/string_writer.hpp>
#include "unpack/auto.hpp"
#include "misc/readjson.hpp"
#include "hash/hash_stream.hpp"
#include <map>
#include <set>

using namespace Tasks;
using namespace SR0;
namespace bs = boost::filesystem;
using namespace Mangle::Stream;
using namespace Mangle::VFS;
using namespace Spread;

typedef std::map<std::string,std::string> StrMap;

/* Converts an archive file into a map of strings. Use the static
   GetArc().
 */
struct Stringify : StreamFactory
{
  StrMap &files;

  Stringify(StrMap &smap) : files(smap) {}

  StreamPtr open(const std::string &name)
  {
    std::string &str = files[name];
    Stream *out = new StringWriter(str);
    return StreamPtr(out);
  }

  static void GetArc(const std::string &arcfile,
                     StrMap &outlist)
  {
    Unpack::AutoUnpack unp;
    unp.unpack(arcfile, StreamFactoryPtr(new Stringify(outlist)));
  }
};

void fail(const std::string &msg)
{
  throw std::runtime_error("SR0 Error: " + msg);
}

#include <iostream>

typedef std::map<Hash, std::string> HashDir;

struct RuleParser;
struct Rule
{
  virtual void index(RuleParser*) = 0;

  /* This is not a usable solution, we need something more akin to the
     job tree we've prototyped elsewhere.

     I can imagine that we have a generalized task collector that you
     provide with input files (and their known hashes), wanted output
     files and their hashes, and then you run that. The output may
     contain duplicate hashes, meaning that the same content should be
     copied into more than one place.

     There is one type specialized for each rule. The tree task system
     is rule- independent, and uses this for the actual tasks. The
     tree system is only responsible for getting necessary
     dependencies, setting up temporary files (if necessary), and
     stringing the tasks together in the right order.
  */
  virtual void writeTo(const std::string &where) = 0;
};

struct RuleParser
{
  typedef std::set<Rule*> RuleSet;
  typedef std::map<Hash, Rule*> RuleMap;

  RuleSet rules;
  RuleMap targets;

  RuleParser() {}

  RuleParser(const Json::Value &val)
  {
    for(int i=0; i<val.size(); i++)
      add(val[i].asString());
  }

  void add(const std::string &rulestr);

  void addTarget(Hash h, Rule *r)
  {
    // TODO: In this very simple test, there's only one rule per
    // target.
    targets[h] = r;
  }
};

struct TestRule : Rule
{
  Hash hash;
  std::string file;

  TestRule(RuleParser *owner, const std::string &str)
  {
    using namespace std;

    int split = str.find(' ');
    if(split == string::npos)
      fail("Invalid TEST rule string '" + str + "'");

    hash.fromString(str.substr(0, split));
    file = str.substr(split+1);

    cout << "Created rule TEST " << hash << " => " << file << endl;
  }

  void index(RuleParser *list)
  {
    list->addTarget(hash, this);
  }

  void writeTo(const std::string &where)
  {
    Tasks::FileOpTask op;
    op.copy(file,where);
    op.run();
    if(op.getInfo()->isNonSuccess())
      fail("Failed to create " + where);
  }
};

void RuleParser::add(const std::string &rulestr)
{
  // The rule 'verb' is the word before the first space
  int pos = rulestr.find(' ');
  std::string verb = rulestr.substr(0, pos);
  std::string rest;
  if(rulestr.size() > verb.size())
    rest = rulestr.substr(pos+1);

  Rule* rule = NULL;
  if(verb == "TEST")
    rule = new TestRule(this, rest);

  rules.insert(rule);
  rule->index(this);
}

static void applySingle(const Json::Value &val, const std::string &file, bs::path dir,
                        RuleParser &rules)
{
  using namespace std;

  if(!val.isMember(file))
    return;

  // Get new file hash
  Hash newHash(val[file].asString());

  // Hash existing file, if any
  Hash oldHash;

  std::string realfile = (dir/file).string();
  if(bs::exists(realfile))
    HashStream::sum(realfile);

  // If it's already up-to-date, exit
  if(oldHash == newHash) return;

  cout << "Updating " << realfile << " from "
       << oldHash << " to " << newHash << endl;

  Rule *r = rules.targets[newHash];
  assert(r);
  r->writeTo(realfile);
}

void SR0::applyFile(const std::string &dir, const std::string &file)
{
  StrMap files;
  Stringify::GetArc(file, files);
  std::string contents = files["index.json"];

  if(contents == "")
    fail("Missing index.json");

  Json::Value val = ReadJson::strToJson(contents);

  RuleParser rules(val["rules"]);

  applySingle(val, "names.json", dir, rules);
  applySingle(val, "rules.json", dir, rules);
}

void SR0::applyURL(const std::string &dir, const std::string &url)
{
  std::string file = (bs::path(dir)/"_sr0_tmp.zip").string();

  InstallTask getFile(url, file, "");
  getFile.run();
  if(!getFile.getInfo()->isSuccess())
    fail("Failed downloading " + url);

  applyFile(dir, file);
  bs::remove(file);
}
