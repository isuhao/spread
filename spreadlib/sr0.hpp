#ifndef __SPREADLIB_SR0_HPP_
#define __SPREADLIB_SR0_HPP_

#include <string>

/* This file is meand to implement the SR0 protocol.

   SR0 is Spread-REST v0, a very early attempt at a common and simple
   protocol for using Spread over HTTP and other transfer protocols.

   No particular care has been taken yet to make sure the protocol
   adheres to the constraints of RESTfulness(*), this more of an
   inspiration and long-term aim.

   (*) http://en.wikipedia.org/wiki/Representational_state_transfer#Constraints
 */

namespace SR0
{
  /* Updates names.json and rules.json in the given directory from the
     given SR0 file.
   */
  void applyFile(const std::string &dir, const std::string &file);

  /* Download the SR0 file from an url, then call applyFile() on it
   */
  void applyURL(const std::string &dir, const std::string &url);
};

#endif
