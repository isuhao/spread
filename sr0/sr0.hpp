#ifndef __SPREAD_SR0_HPP_
#define __SPREAD_SR0_HPP_

/* SR0 (short for Spread-REST v0) is a prototype protocol to install
   or update a directory from data on a web server.

   Unlike the larger Spread system, it does not rely on a complex set
   of rules or pre-existing data. All it needs is an URL (or file
   path) to an sr0/ repository and it will handle the rest.

   SpreadLib uses SR0 to fetch the complete ruleset and data from a
   source.

   SR0 itself uses the Spread rule system to update the directory, and
   thus can much more bandwidth efficient than simply downloading a
   full zip. If set up correctly, knowledge of existing files is used
   to (sometimes greatly) reduce the amount of downloaded data needed.

   SR0 protocol:

   - the SR0 struct takes an input source, either an url or a file
     system directory, depending on whether you use fetchURL() or
     fetchFile()

   - here it expects to find two files:
     - /short.txt
     - /index.zip

   - the short.txt is expected to contain the start of the hash (at
     least 8 bytes of the standard base64url version) of the directory
     that represents the latest version. If this matches the installed
     directory, nothing is done.

   - the existing dir hash is read from /current.hash in the output
     directory, if it exists.

   - if short.txt does NOT match the installed dir (or if there is no
     existing dir/current.hash file), then index.zip is downloaded and
     unpacked to a temporary directory.

   - inside index.zip we expect to find the following files:
     - index.hash - index of the new directory hash. This MUST match
                    the hash in short.txt
     - rules.json - (optional) rules needed to install the dir

     In addition, all files inside the archive are automatically
     hashed and will be used by the installer as needed. You should
     probably at least include the main directory object somewhere in
     the zip.
 */

#include "job/jobinfo.hpp"
#include "cache/cache.hpp"

namespace Spread
{
  struct SR0
  {
    /* Fetch the dir from 'url'.

       The function will check the latest version against a file
       called 'current.hash' in destDir, if it exists. If the dir is
       updated, this file will be updated as well. It's important that
       the data you are downloading does NOT contain any file called
       current.hash, or it will be overwritten!

       The supplied cache should contain any existing files you think
       might help in the process. It will also be updated to index all
       output files.

       If async==true, the install is run in a background thread.
     */
    static JobInfoPtr fetchURL(const std::string &url,
                               const std::string &destDir,
                               Cache::Cache &cache,
                               bool async=true);

    /* Same as fetchURL, but fetch from a filesystem dir instead.

       This may still cause files to be downloaded if the installation
       data itself uses URL rules.
     */
    static JobInfoPtr fetchFile(const std::string &dir,
                                const std::string &destDir,
                                Cache::Cache &cache,
                                bool async=true);
  };
}
#endif
