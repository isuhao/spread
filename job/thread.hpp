#ifndef __JOBIFY_THREAD_HPP
#define __JOBIFY_THREAD_HPP

#include "job.hpp"

namespace Jobify
{
  /* This black-box module swallows any Job object, executes it in a
     separate thread, then deletes the object after it returns. (The
     job must have been allocated using 'new'.)

     You can disable threading (run it locally) by setting
     async=false. This may seem pointless compared to just running
     j->run(), but it does provide a consistent interface for both
     cases, and it will also delete the Job before returning.
   */
  struct Thread
  {
    static void run(Job *j, bool async=true);
    static void sleep(double seconds);
  };
}
#endif
