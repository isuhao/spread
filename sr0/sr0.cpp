#include "sr0.hpp"

#include <boost/filesystem.hpp>
#include <mangle/stream/servers/string_writer.hpp>
#include <mangle/stream/servers/file_stream.hpp>
#include <mangle/stream/servers/outfile_stream.hpp>
#include <mangle/stream/clients/copy_stream.hpp>
#include "htasks/unpackhash.hpp"
#include "job/thread.hpp"
#include "tasks/download.hpp"
#include "rules/rule_loader.hpp"
#include "misc/readjson.hpp"

//#define DEBUG_PRINT
#ifdef DEBUG_PRINT
#include <iostream>
#define PRINT(a) std::cout << __LINE__ << ": " << a << "\n"
#else
#define PRINT(a)
#endif

using namespace Spread;
using namespace Mangle::Stream;
using namespace boost::filesystem;

typedef Hash::DirMap HMap;

struct Sr0Job : Job
{
  std::string source;
  bool isUrl;
  path dest;

  Cache::Cache *cache;
  JobManagerPtr manager;

  bool *wasUpdated;

  // Load current version file
  Hash getVersion(const path &file)
  {
    if(!exists(file)) return Hash();

    uint8_t data[40];

    try
      {
        FileStream inf(file.string());
        if(inf.read(data, 40) != 40)
          return Hash();
      }
    catch(...) { return Hash(); }
    Hash h;
    h.copy(data);
    return h;
  }

  // Load output version from a pack file
  Hash getPackVersion(const path &file)
  {
    if(!exists(file)) return Hash();
    Json::Value val = ReadJson::readJson(file.string());
    return Hash(val["index"]["dirs"][0].asString());
  }

  // Check a short version string against a hash. True if matches.
  bool compVer(const std::string &shortV,
               const Hash &hash)
  {
    // The short version doesn't need to contain the entire hash, just
    // the start of it.
    std::string longV = hash.toString();
    return (shortV.size() >= 8 &&
            shortV.size() <= longV.size() &&
            shortV == longV.substr(0,shortV.size()));
  }

  void doJob()
  {
    PRINT("Starting SR0 job");
    setBusy("Checking for updates");
    assert(cache);
    assert(manager);

    if(wasUpdated) *wasUpdated = false;

    // Load current version indicator, if it exists
    std::string hashFile = (dest/"current.hash").string();
    Hash curHash = getVersion(hashFile);

    // Compare to latest version
    std::string netVer;
    if(!curHash.isNull())
      {
        std::string fetch = source + "/short.txt";

        if(isUrl)
          {
            PRINT("Downloading " << fetch);
            DownloadTask dl(fetch, StringWriter::Open(netVer));
            if(runClient(dl)) return;
            PRINT("   Got: " << netVer);
          }
        else
          {
            CopyStream::copy(FileStream::Open(fetch),
                             StringWriter::Open(netVer));
          }

        // It it a match?
        if(compVer(netVer, curHash))
          {
            // Nothing to do, exit.
            PRINT("No update necessary");
            setDone();
            return;
          }
      }

    setBusy("Fetching update info");
    PRINT("Fetching update info");

    // Next, get the the zip file
    std::string zipfile = source + "/index.zip";
    if(isUrl)
      {
        std::string url = zipfile;
        zipfile = cache->createTmpFilename();
        PRINT("Downloading " << url << " => " << zipfile);
        DownloadTask dl(url, zipfile);
        if(runClient(dl)) return;
        PRINT("  Done.");
      }
    else if(!exists(zipfile))
      {
        setError("File not found: " + zipfile);
        return;
      }

    // Unpack the archive into a temporary directory
    HMap index;
    path tmpOut = cache->createTmpFilename();
    PRINT("Unpacking " << zipfile << " => " << tmpOut);
    UnpackHash::makeIndex(zipfile, index, tmpOut.string());

    // Load the new version
    Hash newHash = getVersion(tmpOut/"index.hash");

    // If there was no index file, try loading the package "index"
    // from packs.json instead
    if(newHash.isNull())
      newHash = getPackVersion(tmpOut/"packs.json");

    if(newHash.isNull() || newHash == curHash ||
       (netVer != "" && !compVer(netVer, newHash)))
      {
        setError("Update version mismatch");
        return;
      }

    // Index all the extracted files
    PRINT("Adding files to cache index");
    for(HMap::iterator it = index.begin(); it != index.end(); it++)
      {
        // Don't use reference, because silly boost paths returns
        // a reference
        std::string file = (tmpOut/it->first).string();
        const Hash &hsh = it->second;
        PRINT(hsh << "  " << file);
        Hash res = cache->index.addFile(file, hsh);
        assert(res == hsh);
      }
    PRINT("  Done.");

    if(checkStatus()) return;

    // Load the ruleset file, if there is one.
    RuleSet rules;
    {
      std::string file = (tmpOut/"rules.json").string();
      if(exists(file))
        loadRulesJsonFile(rules, file);
    }

    InstallerPtr inst = manager->createInstaller(dest.string(), rules);

    /* Add all the hashes in the original zip as archive hints to the
       installer. Since the zip file includes all the necessary dir
       files, some of the hashes in the zip will automatically match
       dirs associated with archives as well. Since we have indexed
       all the files, the Installer should also be able to load all
       the dir objects directly.

       So this essentially preloads all our archives rules. Hashes in
       the zip that do NOT match anything meaningful are just ignored
       by addHint().
     */
    for(HMap::iterator it = index.begin(); it != index.end(); it++)
      inst->addHint(it->second);

    /* Add the requested output directory. Again, since we have
       indexed all the extracted files, the Installer should be able
       to load the dir contents automatically.
    */
    inst->addDir(newHash);

    // Run the install!
    JobInfoPtr inf = manager->addInst(inst);
    setBusy("Updating");
    PRINT("Running main installer => " << dest);
    //assert(inf->isInitiated()); // This may fail
    assert(manager->getInfo()->isInitiated()); // This is correct
    if(waitClient(inf)) return;
    setBusy("Cleaning up");
    PRINT("  Done.");

    // On success, write back the new version file
    OutFileStream out(hashFile);
    out.write(newHash.getData(), 40);

    // Notify the caller that we successfully updated the data, if
    // they requested it.
    if(wasUpdated) *wasUpdated = true;

    setDone();
  }
};

JobInfoPtr SR0::fetchFile(const std::string &dir, const std::string &destDir,
                          JobManagerPtr manager, bool async, bool *wasUpdated)
{
  Sr0Job *j = new Sr0Job;
  j->source = dir;
  j->isUrl = false;
  j->dest = destDir;
  j->cache = &manager->cache;
  j->manager = manager;
  j->wasUpdated = wasUpdated;
  return Thread::run(j,async);
}

JobInfoPtr SR0::fetchURL(const std::string &url, const std::string &destDir,
                         JobManagerPtr manager, bool async, bool *wasUpdated)
{
  PRINT("SR0: Fetching URL " << url << " => " << destDir);

  Sr0Job *j = new Sr0Job;
  j->source = url;
  j->isUrl = true;
  j->dest = destDir;
  j->cache = &manager->cache;
  j->manager = manager;
  j->wasUpdated = wasUpdated;
  return Thread::run(j,async);
}
