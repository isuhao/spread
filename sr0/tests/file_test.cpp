#include "sr0.hpp"

#include <boost/filesystem.hpp>
#include <mangle/stream/servers/outfile_stream.hpp>
#include <dir/binary.hpp>
#include "print_dir.hpp"
#include <job/thread.hpp>

#include <iostream>
using namespace std;

using namespace Mangle::Stream;
using namespace Spread;
namespace bf = boost::filesystem;

// This is used to generate the test set. It is NOT part of the test
// itself.
void genData1()
{
  Hash robo("N0VT8AYfLEu2hFufocgj9ykAQoNEgcQwzLW7m1Tfc-cj");
  Hash out1("psjM8uDq9pmNUZQHCFw8zK4-ayXT5-gpFuksShg-GPUo");
  Hash out2("g2lF9D0PACSgcFZUtHpxhrd6jbdnJmfUyXQ0uwO1VZo-");

  Hash::DirMap dir;
  dir["out_robot.txt"] = robo;
  dir["out_file1.txt"] = out1;
  dir["out_file2.txt"] = out2;
  dir["out_file3.txt"] = out1;

  Hash dirHash = Dir::write(dir, "_outdir1.dat");
  OutFileStream out("_outdir1.hash");
  out.write(dirHash.getData(), 40);
}

void genData2()
{
  Hash out1("psjM8uDq9pmNUZQHCFw8zK4-ayXT5-gpFuksShg-GPUo");
  Hash out2("2-ZutRu-hTIPSAkddcskIPrYFVYK-J9oFSujhDD7bjEn");
  Hash out3("sJjzhl8Vh7vq-hJYYbvYjPKlvPAWc6olCd3Mk_MMv_Qk");
  Hash arcfile("-M2zyJL1Xt9rJYII9fzgh0gS9tHhklFbOTt1wY__Qy9RAg");

  Hash::DirMap adir;
  adir["outfile1.txt"] = out1;
  adir["outfile2.txt"] = out2;
  adir["outfile3.txt"] = out3;

  Hash arcdir = Dir::write(adir, "_arcdir2.dat");

  Hash::DirMap dir;
  dir["out_file1.txt"] = out1;
  dir["out_file2.txt"] = out2;
  dir["out_file4.txt"] = out3;

  Hash dirHash = Dir::write(dir, "_outdir2.dat");
  OutFileStream out("_outdir2.hash");
  out.write(dirHash.getData(), 40);
}

void check(JobInfoPtr info)
{
  if(info->isSuccess())
    cout << "SUCCESS: " << info->getMessage() << " prog=" << info->getCurrent()
         << "/" << info->getTotal() << endl;
  else
    cout << "FAILURE: " << info->getMessage() << endl;
}

void test(const std::string &src, const std::string &dest, bool kill=true)
{
  cout << "Installing " << src << " into " << dest << ":\n";
  Cache::Cache cache;
  JobManagerPtr man(new JobManager(cache));
  cache.tmpDir = "_tmpdir/";
  cache.files.basedir = "_tmpdir/cache/";
  //man->setLogger(Misc::LogPtr(new Misc::Logger()), false);

  if(kill) bf::remove_all(dest);
  bool wasUpdated;

  Thread::run(man);
  check(SR0::fetchFile(src, dest, man, false, &wasUpdated));
  if(wasUpdated) cout << "(UPDATED)\n";
  else cout << "(NOT UPDATED)\n";
  cout << endl;
}

int main()
{
  //genData1();
  //genData2();
  //return 0;

  bf::remove_all("_tmpdir/");

  test("test1", "_outdir1/");
  test("test1", "_outdir2/");
  test("test1", "_outdir2/", false);
  test("test2", "_outdir3");
  test("test1", "_outdir4_1to2");
  test("test2", "_outdir4_1to2", false);
  test("test2", "_outdir5_2to1");
  test("test1", "_outdir5_2to1", false);

  printDir("_outdir1");
  printDir("_outdir2");
  printDir("_outdir3");
  printDir("_outdir4_1to2");
  printDir("_outdir5_2to1");

  /* TODO: if/when we implement dir diffing later, we should recheck
     this example to make sure outdated files are actually deleted
     from the outdir4/5 dirs.
   */

  return 0;
}
