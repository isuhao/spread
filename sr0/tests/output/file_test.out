Installing test1 into _outdir1/:
SUCCESS: Installing prog=35/35

Installing test1 into _outdir2/:
SUCCESS: Installing prog=35/35

Installing test1 into _outdir2/:
SUCCESS: Checking for updates prog=0/0

Installing test2 into _outdir3:
SUCCESS: Installing prog=115/115

Installing test1 into _outdir4_1to2:
SUCCESS: Installing prog=35/35

Installing test2 into _outdir4_1to2:
SUCCESS: Installing prog=75/75

Installing test2 into _outdir5_2to1:
SUCCESS: Installing prog=115/115

Installing test1 into _outdir5_2to1:
SUCCESS: Installing prog=35/35

